plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XWorkspaceManager.java

```diff
@@ -428,14 +429,46 @@ public XProjectManager getProjectManager(String projectName) {
 	}
 
 	/**
-	 * Mark the given document as cloded.
+	 * Mark the given document as closed.
+	 *
+	 * When we close a document, being it a saved document or a dirty document, we need to run a build to make sure that
+	 * the generated code is consistent.
+	 *
+	 * Reason:<br>
+	 * Two files are open, dirty and refer to each other. One file is saved and code is generated. This generation is
+	 * done against the dirty (!) state of the other open document in LSP. When the other file is closed, the generator
+	 * needs to run again.
```

Comment:
Somehow I wonder how to make the implementation of the constraint "Source and transpiled files are always in sync on disk" more obvious. Maybe even including the n4js.projectstate. (Whereas the internal type index always relates to the current editor state, no matter of being in sync with the file on disk or being dirty instead.) ... Future refactorings.

Reply:
👍